### PR TITLE
choose which function should be executed

### PR DESCRIPTION
**Pull request by Geal**
it will either use the start function (if provided), or an exported
function, indicated by its name.

Note: it assumes that the start function or the exported ones all have the same prototype. To get more interesting behaviour, we should probably move the preparation of the context ( https://github.com/CraneStation/wasmtime/blob/master/lib/execute/src/execute.rs#L125-L154 ) to another function that could be called directly.

### CODE REVIEW

**Comment by sunfishcode** [comment]
In the case where we call an export, should we implicitly call the start function first? A wasm module might be relying on the start function to initialize itself before any of its exports are called.

**Comment by Geal** [comment]
is it something that all `$language -> wasm` implementations do, or only the Rust ones?
I can add it, it's not too hard.
Should I make a PR to separate the execution preparation into another function?

**Comment by sunfishcode** [comment]
This isn't a Rust-specific issue; it's a wasm one. Consulting the spec, calling the start function is considered to be part of module instantiation. So yeah, if you want to separate execution preparation into a separate function, that'd be good.

**Comment by Geal** [comment]
what if we want to call the exported function multiple times? We should not call the start function every time, right?
Here's something I could do:
- move the execution preparation into another function, to set up code and memory
- make a `start` function that will call the module's `start` if it's there
- make another function (`execute`? `execute_export`?) that would be used to call exported functions. Possibly with a bool somewhere in the `Instance` to make sure that `start` was called
By the way, how could I pass arguments to exported functions?

**Comment by sunfishcode** [comment]
Yes, that sounds right.

For passing in arguments, that's... an interesting question :-). To start, it might be a good time to introduce an `enum Value` which can hold an i32, i64, f32, or f64 value, and then we can pass something like a `&[Value]` along with the export function to call. But then, we need to construct the call. Since we have a JIT handy, I wonder if the best way to do this is just to JIT a little function on the spot that contains the arguments as constants and makes the call.

**Comment by Geal** [comment]
not only for arguments, but also for the return value. That part should be interesting, but first I'll make the functions I mentioned, and assume the exported functions have the same prototype as `start`

**Comment by Geal** [comment]
@sunfishcode alright, here's a new version. I'm not too happy about it yet. I think the call to `start` should be part of the `prepare_execute` function. This would avoid that `initialized` flag that may or may not be true when we call `execute` afterwards.
Also, a side question: how is the `vmctx` handled on the wasm side in the generated code? Is there a function prefix that sets up the vmctx somewhere on the stack?

File: lib/execute/src/execute.rs [hunk 1]
```diff
@@ -154,26 +158,73 @@ pub fn execute(
...
+    compilation: &Compilation,
+    context: &mut Context,
+) -> Result<(), String> {
+    if context.initialized {
+        return Err(String::from("the start function was already called"));
+    }
+
+    let start_index = module
+        .start_func
+        .ok_or_else(|| String::from("No start function defined, aborting execution"))?;
```
Comment (sunfishcode) [inline lib/execute/src/execute.rs]: If there's no start function defined, we should just return here, without reporting an error.

**Comment by sunfishcode** [comment]
> @sunfishcode alright, here's a new version. I'm not too happy about it yet. I think the call to `start` should be part of the `prepare_execute` function. This would avoid that `initialized` flag that may or may not be true when we call `execute` afterwards.

Oh, heh. I missed that you said this. I had the same thoughts as I reviewed the patch :-).

> Also, a side question: how is the `vmctx` handled on the wasm side in the generated code? Is there a function prefix that sets up the vmctx somewhere on the stack?

The vmctx is allocated in `make_vmctx`, and then the address of the buffer is passed as an extra argument to all functions.

**Comment by Geal** [comment]
I just pushed the version with the `finish_instantiation` idea.
I think it would make sense to perform the initialization inside `Instance::new()`. There could then be a `Instance::execute()` that only takes the function name (and its arguments, once we get to it), provided I add to `Instance` members for `&Module`, `&Compilation` and `vmctx`.

**Comment by Geal** [comment]
and here's another commit for the version where everything is moved in `Instance`. I thnik it looks much better this way :)

**Comment by Geal** [comment]
it makes usage a bit harder though: when imported functions are called, they might want to access the `Instance` to read or write the memory. In my tests, I was putting it in TLS, but with this design, `Instance` now comes with a lifetime that's not very compatible with TLS.
Also, calling `execute()` now means borrowing `Instance` mutably, but when we get into the imported function, we might want to borrow it mutably again.

**Comment by Geal** [comment]
I think I'll wait for #22 to be merged, and then rebase this one on it then

**Comment by Geal** [comment]
@sunfishcode I see that #22 did not port the execute crate to using VMContext, I can give it a go if you'd like

**Comment by sunfishcode** [comment]
@Geal Yes, that's the next step there :-). If you want to look at that, I'll start working on trap handling.

**Comment by sunfishcode** [comment]
Looks good. Let's merge this, and we can build from here!

**Pull request merged**